cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -848,11 +877,12 @@ public boolean apply(RunRecordMeta runRecordMeta) {
   /**
    * Upgrades the keys in table to include version if it is not already present.
    */
-  void upgradeVersionKeys() {
-    upgradeVersionKeys(TYPE_APP_META, ApplicationMeta.class);
-    upgradeVersionKeys(TYPE_RUN_RECORD_COMPLETED, RunRecordMeta.class);
-    upgradeVersionKeys(TYPE_WORKFLOW_NODE_STATE, WorkflowNodeStateDetail.class);
-    upgradeVersionKeys(TYPE_WORKFLOW_TOKEN, BasicWorkflowToken.class);
+  boolean upgradeVersionKeys(int maxRows) {
+    boolean upgradeDone = upgradeVersionKeys(TYPE_APP_META, ApplicationMeta.class, maxRows);
+    upgradeDone &= upgradeVersionKeys(TYPE_RUN_RECORD_COMPLETED, RunRecordMeta.class, maxRows);
+    upgradeDone &= upgradeVersionKeys(TYPE_WORKFLOW_NODE_STATE, WorkflowNodeStateDetail.class, maxRows);
+    upgradeDone &= upgradeVersionKeys(TYPE_WORKFLOW_TOKEN, BasicWorkflowToken.class, maxRows);
+    return upgradeDone;
```

Comment:
`this.upgradeComplete` is only updated in `DefaultStore.DefaultStoreUpgradeCacheLoader.load(byte[] key)`, otherwise, it's not updated. It can actually be updated here if `upgradeDone` is true. If this is done, maybe `DefaultStore.DefaultStoreUpgradeCacheLoader` won't be necessary and `DefaultStore.isUpgradeComplete()` can just return  `upgradeComplete` in AppMetadataStore